query-engine/core/src/telemetry/capturing/capturer.rs

```diff
@@ -156,70 +112,199 @@ impl SpanProcessor for Processor {
-        self.value
-            .attributes
-            .retain(|key, _| (&VALID_QUERY_ATTRS).contains(&key.as_str()));
+        fn query_event(mut self) -> models::LogEvent {
+            self.value
+                .attributes
+                .retain(|key, _| (&VALID_QUERY_ATTRS).contains(&key.as_str()));
 
-        models::LogEvent {
-            level: "query".to_string(),
-            ..self.value
+            models::LogEvent {
+                level: "query".to_string(),
+                ..self.value
+            }
         }
-    }
 
-    #[inline(always)]
-    fn is_loggable_event(&self) -> bool {
-        self.settings.included_log_levels.contains(&self.value.level)
+        #[inline(always)]
+        fn is_loggable_event(&self) -> bool {
+            self.settings.included_log_levels.contains(&self.value.level)
+        }
     }
-}
-
-/// tests for capture exporter
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use std::time::Duration;
 
-    #[tokio::test]
-    async fn test_garbage_collection() {
-        let exporter = Processor::new();
+    #[derive(Debug)]
+    pub(super) enum CaptureOp {
+        /// Tells the task that the given span data has been processed by the span processor
+        SpanDataProcessed(SpanData),
+        /// Tells the task to start capturing for the given trace id
+        StartCapturing(TraceId, Settings, oneshot::Sender<()>),
+        /// Tells the task to fetch the captures for the given trace_id, and sendthem to the given sender
+        FetchCaptures(TraceId, oneshot::Sender<Storage>),
+        /// Tells the task to cleaup traces that could have potentially expired
+        GC(TraceId),
+    }
 
-        let trace_id = TraceId::from_hex("1").unwrap();
-        let one_ms = Duration::from_millis(1);
+    static TASK: Lazy<Sender<CaptureOp>> = Lazy::new(|| {
+        let (sender, receiver) = unbounded();
+        let cleanup_sender = sender.clone();
+
+        std::thread::spawn(move || {
+            let mut store: HashMap<TraceId, Storage> = Default::default();
+
+            let tokio = tokio::runtime::Builder::new_multi_thread()
+                .enable_all()
+                .build()
+                .unwrap();
+
+            loop {
+                match receiver.recv() {
+                    Ok(CaptureOp::StartCapturing(trace_id, settings, op_sender)) => {
+                        tracing::trace!("capture task: start capturing for {:?}", trace_id);
+
+                        let storage = Storage::from(settings);
+                        let ttl = storage.settings.ttl;
+                        store.insert(trace_id, storage);
+                        _ = op_sender.send(());
+
+                        let cleanup_sender = cleanup_sender.clone();
+                        tokio.spawn(async move {
+                            tokio::time::sleep(ttl).await;
+                            cleanup_sender.send(CaptureOp::GC(trace_id)).unwrap();
+                        });
```

Comment:
This is the most controversial part of this PR. I had to instantiate a reactor in order to spawn tasks at this level. If this was not in place, the tokio runtime was not set, which was unexpected to me. Given the evaluation of this code happens after the server was initiated, which is decorated with the `#[tokio::main]` macro.

Reply:
Yes, it probably works in many circumstances but from experience, starting two tokio multi-threaded tokio runtimes per query engine process is going to bring us a lot of pain.

I see two ways out of this:

1. Spawn a long-lived tokio task, use tokio channels instead of spawning a thread and using crossbeam channels.
2. Make all the code in the thread sync, do not involve tokio at all. I'd need to re-read and re-understand the whole thing to see if I am missing anything that needs to be async but I don't think there is. The garbage collection would have to work differently.

From gut feeling, (1.) should be less effort from where the code is now and work for what you are trying to achieve.